File: examples/buttons/server/rails/app/models/shop_order.rb
# frozen_string_literal: true

# One cart checkout. The id IS the OpenReceive `reference`: created before
# checkout, kept across every retry, never reused, and unguessable because it is
# a uuid rather than a sequential integer.
#
# OpenReceive never sees this model. The three hooks in
# config/initializers/openreceive.rb are the only bridge:
#   authorize   -> shop_user_id vs. the signed cookie
#   amount_for  -> total_amount / checkout_description, below
#   on_paid     -> claim_paid!, below
class ShopOrder < ApplicationRecord
  AWAITING_PAYMENT = "awaiting_payment"
  PAID = "paid"
  STATES = [AWAITING_PAYMENT, PAID].freeze

  UUID_PATTERN = /\A\h{8}-\h{4}-\h{4}-\h{4}-\h{12}\z/

  # How many rows the public feed shows, and the only limit it honours.
  FEED_LIMIT = 25

  belongs_to :shop_user, inverse_of: :shop_orders

  has_many :items, -> { order(:created_at) },
           class_name: "ShopOrderItem", dependent: :destroy, inverse_of: :shop_order

  validates :state, inclusion: { in: STATES }
  validates :total_cents, numericality: { only_integer: true, greater_than: 0 }

  scope :paid, -> { where(state: PAID) }
  # Newest paid first, with users and items preloaded — the feed is an N+1 by
  # default and this is a demo people read.
  scope :recent, -> { paid.order(paid_at: :desc, created_at: :desc) }

  # The reference arrives as a string the payer's browser sent. Postgres RAISES
  # on a malformed uuid, so the format check happens before the query — every
  # caller here is on a request path an anonymous payer can reach.
  def self.find_by_reference(reference)
    return nil unless reference.is_a?(String) && reference.match?(UUID_PATTERN)

    find_by(id: reference)
  end

  # One transaction: the order and every item, with the totals summed from the
  # PRODUCT rows the controller looked up. Nothing here reads a number the
  # browser sent.
  #
  # The name and unit price are copied onto the item deliberately — see
  # ShopOrderItem for why history must not move when the catalog does.
  def self.create_from_lines!(lines, shop_user:)
    transaction do
      order = create!(
        shop_user: shop_user,
        state: AWAITING_PAYMENT,
        currency: "USD",
        total_cents: lines.sum { |line| line.fetch(:product).price_cents * line.fetch(:quantity) }
      )

      lines.each do |line|
        product = line.fetch(:product)
        order.items.create!(
          product: product,
          sku: product.sku,
          name: product.name,
          unit_price_cents: product.price_cents,
          quantity: line.fetch(:quantity)
        )
      end

      order
    end
  end

  def paid?
    state == PAID
  end

  # The settlement push, on two streams: this payer's own order, and the public
  # feed every visitor watching the Recent orders tab is subscribed to.
  #
  # NEVER call this from `config.on_paid`. That hook runs INSIDE the settlement
  # transaction, where the rule is database writes only — a broadcast sent from
  # there would survive a rollback and tell every browser about an order that
  # does not exist. The initializer schedules it with
  # `ActiveRecord.after_all_transactions_commit` instead, and there is a test
  # that rolls the transaction back and asserts silence.
  #
  # Best-effort by construction: a dropped broadcast costs latency, never
  # correctness, because both the feed and the checkout keep polling.
  def broadcast_settled!
    ShopOrderChannel.broadcast_paid(id)
    ShopFeedChannel.broadcast_orders_changed
  end

  # A decimal string for `config.amount_for`. Integers all the way down; the
  # division happens once, in the formatter. Never a float in a payload.
  def total_amount
    format("%.2f", total_cents / 100.0)
  end

  # What the payer is BUYING, in our own words — one display string, rendered
  # above the amount on every checkout screen. Without it the checkout is a QR
  # code and "$4.00" with no sign of what the four dollars is for, because
  # OpenReceive owns no line items and can show nothing else on its own.
  #
  # Built from the item SNAPSHOTS, so it reads the same after a catalog edit.
  def checkout_description
    parts = items.map do |item|
      label = item.name.presence || item.sku
      item.quantity > 1 ? "#{label} ×#{item.quantity}" : label
    end

    "OpenReceive #{'button'.pluralize(items.sum(&:quantity))}: #{parts.join(', ')}"
  end

  # THE GUARDED TRANSITION, idempotent by construction: the WHERE clause is the
  # lock. Whoever flips awaiting_payment -> paid first is the only one who
  # fulfills; a later attempt updates zero rows and does nothing. Returns true
  # when THIS call is the one that claimed the order.
  #
  # A class method, not an instance method, and `update_all` rather than
  # `update!`: one conditional UPDATE with no model code and no callbacks
  # between the check and the write. There is nowhere for a second caller to
  # squeeze in.
  #
  # OpenReceive already guarantees the settlement hook runs at most once per
  # reference across every path it owns. This is still written this way because
  # OpenReceive cannot see a second fulfillment path of OURS — an admin action,
  # a support tool, a replayed job — and the moment one exists, those race each
  # other rather than the library.
  def self.claim_paid!(reference:, paid_at:, payment_hash:)
    return false unless reference.is_a?(String) && reference.match?(UUID_PATTERN)

    claimed = where(id: reference, state: AWAITING_PAYMENT).update_all(
      state: PAID,
      paid_at: paid_at,
      payment_hash: payment_hash,
      updated_at: Time.current
    )

    claimed.positive?
  end
end

File: examples/buttons/server/laravel/app/Models/ShopOrder.php
<?php

declare(strict_types=1);

namespace App\Models;

use App\Support\Uuid;
use Carbon\CarbonImmutable;
use Illuminate\Database\Eloquent\Concerns\HasUuids;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;
use Illuminate\Support\Facades\DB;

/**
 * One cart checkout. The id IS the OpenReceive `reference`: created before
 * checkout, kept across every retry, never reused, and unguessable because it
 * is a uuid rather than a sequential integer.
 *
 * OpenReceive never sees this model. The three hooks in app/OpenReceive/Host.php
 * are the only bridge:
 *   authorize  -> shop_user_id vs. the encrypted visitor cookie
 *   amountFor  -> totalAmount() / checkoutDescription(), below
 *   onPaid     -> claimPaid(), below
 */
class ShopOrder extends Model
{
    use HasUuids;

    public const AWAITING_PAYMENT = 'awaiting_payment';
    public const PAID = 'paid';

    /** How many rows the public feed shows, and the only limit it honours. */
    public const FEED_LIMIT = 25;

    protected $guarded = [];

    protected $casts = [
        'total_cents' => 'integer',
        'paid_at' => 'datetime',
    ];

    public function shopUser(): BelongsTo
    {
        return $this->belongsTo(ShopUser::class, 'shop_user_id');
    }

    public function items(): HasMany
    {
        // Catalog order: items are created in the order the cart was normalized
        // (position, price), and HasUuids mints time-ordered ids.
        return $this->hasMany(ShopOrderItem::class, 'shop_order_id')->orderBy('created_at')->orderBy('id');
    }

    /** The reference arrives as a string the payer's browser sent: Postgres RAISES on a malformed uuid literal, so it is checked first. */
    public static function findByReference(mixed $reference): ?self
    {
        return Uuid::valid($reference) ? self::query()->find($reference) : null;
    }

    /**
     * One transaction: the order and every item, with the total summed from
     * the PRODUCT rows the controller looked up. Nothing here reads a number
     * the browser sent. Name and unit price are snapshotted onto the item so a
     * receipt survives a renamed or deleted product.
     *
     * @param list<array{product: ShopProduct, quantity: int}> $lines
     */
    public static function createFromLines(array $lines, ShopUser $user): self
    {
        return DB::transaction(static function () use ($lines, $user): self {
            $order = self::query()->create([
                'shop_user_id' => $user->id,
                'state' => self::AWAITING_PAYMENT,
                'currency' => 'USD',
                'total_cents' => array_sum(array_map(static fn (array $line): int => $line['product']->price_cents * $line['quantity'], $lines)),
            ]);
            foreach ($lines as $line) {
                $order->items()->create([
                    'product_id' => $line['product']->id,
                    'sku' => $line['product']->sku,
                    'name' => $line['product']->name,
                    'unit_price_cents' => $line['product']->price_cents,
                    'quantity' => $line['quantity'],
                ]);
            }
            return $order;
        });
    }

    public function isPaid(): bool
    {
        return $this->state === self::PAID;
    }

    /** A decimal STRING from integer cents. Never a float: the division happens once, here, at the edge. */
    public function totalAmount(): string
    {
        return sprintf('%d.%02d', intdiv($this->total_cents, 100), $this->total_cents % 100);
    }

    /** "OpenReceive buttons: Safety Orange ×2, Classic Black" — what `amountFor` shows the payer above the amount. */
    public function checkoutDescription(): string
    {
        $items = $this->items;
        $count = (int) $items->sum('quantity');
        $parts = $items->map(static fn (ShopOrderItem $item): string => $item->quantity > 1 ? "{$item->name} ×{$item->quantity}" : $item->name);
        return 'OpenReceive '.($count === 1 ? 'button' : 'buttons').': '.$parts->implode(', ');
    }

    /**
     * THE GUARDED TRANSITION THE MONEY RESTS ON. One conditional UPDATE: the
     * WHERE clause is the lock, so whoever flips awaiting_payment -> paid first
     * is the only one who fulfills, and every later caller updates 0 rows.
     * Runs on the connection OpenReceive's settlement transaction holds.
     */
    public static function claimPaid(string $reference, int $paidAt, string $paymentHash): bool
    {
        if (!Uuid::valid($reference)) {
            return false;
        }
        $claimed = self::query()
            ->where('id', $reference)
            ->where('state', self::AWAITING_PAYMENT)
            ->update([
                'state' => self::PAID,
                'paid_at' => CarbonImmutable::createFromTimestampUTC($paidAt),
                'payment_hash' => $paymentHash,
                'updated_at' => now(),
            ]);
        return $claimed > 0;
    }
}
